File: leetcode/363/c_MaxNumAlloysⅠ.cpp
#include <bits/stdc++.h>

using namespace std;
class Solution {
public:
    int maxNumberOfAlloys(int n, int k, int budget, vector<vector<int>>& composition, vector<int>& stock, vector<int>& cost) {
        auto check = [&](long long X) {
            for(auto &cmp : composition)
            {
                long long money = 0;
                for(int i=0;i<n;++i)
                {
                    if(stock[i] < X*cmp[i])
                        money += (X*cmp[i]-stock[i])*cost[i];
                }
                if(money <= budget) return true;
            }
            return false;
        };

        long long l= 0, r = 2e8;
        while(l < r)
        {
            long long mid = l +  ((r - l + 1) >> 1);
            if(check(mid)) l = mid;
            else r = mid - 1;
        }
        return l;
    }
};

File: _LeetCode/363/c_MaxNumAlloysⅠ.py
from typing import List
class Solution:
    def maxNumberOfAlloys(self, n: int, k: int, budget: int, composition: List[List[int]], stock: List[int], cost: List[int]) -> int:
        ans = 0
        mx = min(stock) + budget
        for com in composition:
            def check(num: int) -> int:
                money = 0
                for s, base, c in zip(stock, com, cost):
                    if s < base * num:
                        money += (base * num - s) * c
                        if money > budget:
                            return False
                return True

            left, right = 0, mx + 1
            while left + 1 < right:  # 开区间写法
                mid = (left + right) // 2
                if check(mid):
                    left = mid
                else:
                    right = mid
            ans = max(ans, left)
        return ans

if __name__ == "__main__":
    n = 3
    k = 2
    budget = 0
    composition = [[2,2,2],[1,10,1]]
    stock = [3,12,3]
    cost = [10,10,10]
    print(Solution().maxNumberOfAlloys(n,k,budget,composition,stock,cost))